File: 프로그래머스/2/42584. 주식가격/주식가격.java
import java.util.*;

class Solution {
    public int[] solution(int[] prices) {
        int[] answer = new int[prices.length];
        Stack<ArrayList<Integer>> q = new Stack<ArrayList<Integer>>();
        for(int i=0;i<prices.length;i++){
            if(!q.isEmpty()){
                while(!q.isEmpty()){
                    ArrayList<Integer> list = q.peek();
                    if(list.get(0)>prices[i]){
                        q.pop();
                        answer[list.get(1)] = i-list.get(1);
                    }else break;
                }
            }
            q.add(new ArrayList<>(Arrays.asList(prices[i],i)));
        }
        while(!q.isEmpty()){
            ArrayList<Integer> list = q.pop();
            answer[list.get(1)] = prices.length - list.get(1)-1;
        }
        return answer;
    }
}

File: 프로그래머스/2/42584. 주식가격/주식가격.js
function solution(prices) {
    var answer = Array.from({length:prices.length},_=>0);
    const con = new Map()
    
    for(let i=0;i<prices.length;i++){
        let end = -1
        //console.log(con)
        if(con.get(prices[i])<=i||con.get(prices[i])===undefined){
            for(let j=i+1;j<prices.length;j++){
                if(prices[j]<prices[i]){
                    end = j
                    break;
                }
            }
        }else end = con.get(prices[i])
        if(end===-1){
            answer[i] = prices.length - i - 1
        }else answer[i] = end - i
        
        con.set(prices[i],end)
    }
    return answer;
}